File: heron/dsl/src/java/com/twitter/heron/dsl/Context.java
//  Copyright 2017 Twitter. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.twitter.heron.dsl;


import java.util.HashSet;
import java.util.Set;

import com.twitter.heron.api.Config;
import com.twitter.heron.api.HeronSubmitter;
import com.twitter.heron.api.exception.AlreadyAliveException;
import com.twitter.heron.api.exception.InvalidTopologyException;
import com.twitter.heron.api.topology.TopologyBuilder;

/**
 * A Streamlet is a (potentially unbounded) ordered collection of tuples.
 Streamlets originate from pub/sub systems(such Pulsar/Kafka), or from static data(such as
 csv files, HDFS files), or for that matter any other source. They are also created by
 transforming existing Streamlets using operations such as map/flatMap, etc.
 Besides the tuples, a Streamlet has the following properties associated with it
 a) name. User assigned or system generated name to refer the streamlet
 b) nPartitions. Number of partitions that the streamlet is composed of. The nPartitions
 could be assigned by the user or computed by the system
 */
public final class Context {
  public static void run(String name, Config config, Builder builder) {
    Set<String> stageNames = new HashSet<>();
    TopologyBuilder topologyBuilder = builder.build();
    try {
      HeronSubmitter.submitTopology(name, config, topologyBuilder.createTopology());
    } catch (AlreadyAliveException e) {
      e.printStackTrace();
    } catch (InvalidTopologyException e) {
      e.printStackTrace();
    }
  }

  private Context() { }
}

File: heronpy/streamlet/context.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

'''context.py: module for defining context'''

from abc import abstractmethod

class Context:
  """Context is the information available at runtime for operators like transform.
     It contains basic things like config, runtime information like task,
     the stream that it is operating on, ProcessState, etc.
  """

  @abstractmethod
  def get_task_id(self):
    """Fetches the task id of the current instance of the operator
    """

  @abstractmethod
  def get_config(self):
    """Fetches the config of the computation
    """

  @abstractmethod
  def get_stream_name(self):
    """Fetches the stream name that we are operating on
    """

  @abstractmethod
  def get_num_partitions(self):
    """Fetches the number of partitions of the stream we are operating on
    """

  def get_partition_index(self):
    """Fetches the partition of the stream that we are operating on
    """

  @abstractmethod
  def get_state(self):
    """The state where components can store any of their local state
    """

  @abstractmethod
  def emit(self, values):
    """Emits the values in the output stream
    """
